### PR TITLE
Bazel 0.16.1, `Tulsi` bump, `rules_apple` bump

### PR DESCRIPTION
**Pull request by jerrymarino**
This patch updates XCHammer for Bazel 0.16.1 addressing changes since 0.12.

It includes an update to the entitlements export helper for `rules_apple` changes.

Additionally flip on Bazel targets, by default. Mainly, this makes testing easier. Perhaps this will be split to a separate commit / turned on in XCHammer internally.

### CODE REVIEW

**Comment by jerrymarino** [comment]
The current revision of this PR gets XCHammer in a state where Xcode projects are generating, and building with Xcode. The Bazel build is broken - output from Bazel is not updated in the log anymore. 

XCHammer currently runs Bazel with a build script, `build_bazel.py`, forked from the `Tulsi` project which is where the problem lies.There have been changes with the way that output is read from Bazel since the last base commit ( nearly 1 year old! ) and I'm experimenting to see if bumping Tulsi can fix the issue.

The goal of the commit: https://github.com/pinterest/tulsi/commit/dcc0b0abed06e77a3f8f41b4bbbaa1c040f279ac is to collect all ad-hoc changes to Tulsi, and then propose them to the Tulsi team, killing the Pinterest fork 💵 

In the worst case, it'll be somewhat easier to reason about how the XCHammer fork diverges.

File: sample/UrlGet/UrlGet.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -44,6 +44,7 @@
 		BF_158572982416 /* Tests2.m in Sources */ = {isa = PBXBuildFile; fileRef = FR_447194245190 /* Tests2.m */; };
 		BF_161334189326 /* STPEmailAddressValidator.m in Sources */ = {isa = PBXBuildFile; fileRef = FR_820602908988 /* STPEmailAddressValidator.m */; };
 		BF_164244000470 /* GoogleAppIndexing.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = FR_353990188355 /* GoogleAppIndexing.framework */; };
+		"BF_164244000470-1" /* GoogleAppIndexing.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = FR_353990188355 /* GoogleAppIndexing.framework */; };
```
Comment (keith) [inline sample/UrlGet/UrlGet.xcodeproj/project.pbxproj]: This seems unintentional

Comment (jerrymarino) [reply sample/UrlGet/UrlGet.xcodeproj/project.pbxproj]: Hey Keith! You're right - I'm not sure what exactly caused this to pop up. It looks like now, `.framework` is included 2x in the `Link with libraries` phase for the Xcode built targets. My intention is to resolve this as part of wrapping up the upgrade ( and soon ) 😄

**Comment by jerrymarino** [comment]
I updated mostly everything and addressed all known issues. Integration tests / manual tests of the sample check out. Planning to spend a few more cycles tomorrow AM testing it out before merging!

This patch updates XCHammer for Bazel 0.16.1

- `rules_apple` updates

    Bump sample to latest `rules_apple` - `Tulsi` is running on HEAD of `rules_apple` today.

    Additionally, includes an update to the entitlements export helper for `rules_apple` changes.

- `Tulsi` dependency simplification

    `Tulsi` is updated to `HEAD` of `Tulsi` which corresponds to this Bazel version.  After a year of updates, all XCHammer `Tulsi` patches are consolidated and will be proposed to the `Tulsi` team.

    This patch reworks the packaging structure to be more aligned with how NSBundle / Apple expects a package to work. The package structure is now flat: tulsi-aspects is smashed into the root.  This removes the need to have several nasty hacks to the `Tulsi` code directly.

    Import `TulsiRuleEntryMapExtractor`. Previously, this was added into `Tulsi`'s source code directly. By making it's dependencies public, this can live in XCHammer.

- Bazel Build Infrastructure changes

    This change includes implementing `BazelBuildSettings` per the latest `Tulsi`. `bazel_build.py` reads the command line arguments for Bazel from a code generated python program, `bazel_build_settings.py`.  This patch reuses this pattern by reusing the corresponding `Tulsi` types.  An `XCHammerAsset` is codegen'd into each project. `bazelExecRoot` / `repository_override`  are now handled by XCHammer.

- Misc changes

    Flip on Bazel targets, by default. Mainly, this makes testing easier and is the direction XCHammer is headed. Perhaps there will be another commit to turn them on in XCHammer internally.

    Add clarity and expectations around versioning of XCHammer.

File: Sources/XCHammer/XcodeTarget.swift [hunk 2]
```diff
@@ -989,11 +976,19 @@ public class XcodeTarget: Hashable, Equatable {
...
                 return """
                 export TULSI_USE_HAMMER_DEBUG_CONFIG=YES
+                export PYTHONPATH=\"$PROJECT_FILE_PATH/XCHammerAssets\"
                 \(buildInvocation)
                 """
             }
+
+            // Note: when using a custom template, the user is required to
+            // insert:
+            // export PYTHONPATH=\"$PROJECT_FILE_PATH/XCHammerAssets\"
```
[the remark this answers is not in the document]
Comment (jerrymarino) [reply Sources/XCHammer/XcodeTarget.swift]: We can probably write this as a build setting into the Xcode proj.

File: Sources/XCHammer/XcodeTarget.swift [hunk 3]
```diff
@@ -961,26 +964,10 @@ public class XcodeTarget: Hashable, Equatable {
...
-           cmd = cmd + ["--bazel_options", buildBazelOptions, "--"]
-        }
-
-        if let buildBazelStartupOptions = targetConfig?.buildBazelStartupOptions {
-           cmd = cmd + ["--bazel_startup_options", buildBazelStartupOptions, "--"]
-        }
-
-        // tulsi-aspects are adjacent to the XCHammer bin
+        // bazel_build.py is adjacent to the XCHammer bin
+        // TODO: what are the actual arguments we need to use
```
[the remark this answers is not in the document]
Comment (rahul-malik) [reply Sources/XCHammer/XcodeTarget.swift]: Maybe remove this TODO if its' already out-dated before it's merged?

File: sample/UrlGet/UrlGet.xcodeproj/XCHammerAssets/bazel_build_settings.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    buildFlags.extend(cache_affecting.build)
+    buildFlags.extend(lang.build)
+
+    return (self.bazel, startupFlags, buildFlags)
+
+# Default value in case the template does not behave as expected.
+BUILD_SETTINGS = None
+
+BUILD_SETTINGS = BazelBuildSettings(
+    '/Users/jerry/Projects/xchammer-github/sample/UrlGet/tools/bazelwrapper',
```
Comment (rahul-malik) [inline sample/UrlGet/UrlGet.xcodeproj/XCHammerAssets/bazel_build_settings.py]: Absolute paths?

Comment (jerrymarino) [reply sample/UrlGet/UrlGet.xcodeproj/XCHammerAssets/bazel_build_settings.py]: Yep - this is a code generated asset under the Xcode proj. If we want to do remote caching of `xcodeproj`, this can't be part of that cache, or the `execRoot` will have to be determined at runtime.

File: Sources/XCHammer/XcodeTarget.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerrymarino) [reply Sources/XCHammer/XcodeTarget.swift]: 💯

**Comment by jerrymarino** [comment]
I'm pretty happy with this change overall. I've spent some time testing it with remote caching which unveiled that the default options are quite different now.
Default setting changes:
- it uses `cpu` instead of `config=`. For me, I need the robustness of a `config` for a `cpu`
- it sets up `watchos` cpus
- it passes `'--apple_platform_type=ios'`
- `compilation_mode` is not set anymore - not sure if `XCHammer` should do something sensible.

Additionally
- it looks like the bazel action `build` no longer cache hits with the action `test`. If this is verified true, this will need to be some CLI flag so that way I can set the action to `test` ( if you use `test` to write actions to the cache then it won't work ).
- with the addition of `BazelBuildSettings`, it is not very easy to point it at shell variables anymore. I needed to do more hacks to `bazelwrapper` to enable this, which should be added to the sample. This is a blocker for static analysis and remote caching of such static analysis in Xcode.

In summary, this PR is pretty big already - and it needs more work for remote caching / my usage.

I'll submit a follow up that:
- allows the user to override `BazelBuildSettings.platformConfigurationFlags`
- allows the user to specify the bazel action.
